exposed-tests/src/test/kotlin/org/jetbrains/exposed/sql/tests/shared/dml/MergeTableTest.kt

```diff
@@ -0,0 +1,489 @@
+
+            Insert.dest("test1", 100)
+            Insert.dest("test2", 100)
+            Insert.dest("test3", 100)
+
+            Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                deleteWhenMatched(and = (Source.value less 7))
+                deleteWhenMatched(and = (Source.value greater 12))
+            }
+
+            assertNull(Dest.getByKeyOrNull("test1"))
+            assertNotNull(Dest.getByKeyOrNull("test2"))
+            assertNull(Dest.getByKeyOrNull("test3"))
+        }
+    }
+
+    @Test
+    fun testMultipleInsertsInOracleAndSqlserver() {
+        withMergeTestsTables(excludeSettings = allDbExcept(TestDB.oracleRelatedDB + TestDB.sqlServerRelatedDB)) {
+            assertFailsWith(IllegalArgumentException::class) {
+                Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                    insertWhenNotMatched {
+                        it[Dest.key] = Source.key
+                        it[Dest.value] = Source.value
+                    }
+                    insertWhenNotMatched {
+                        it[Dest.key] = Source.key
+                        it[Dest.value] = Source.value
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun testMultipleUpdatesInOracleAndSqlserver() {
+        withMergeTestsTables(excludeSettings = allDbExcept(TestDB.oracleRelatedDB + TestDB.sqlServerRelatedDB)) {
+            assertFailsWith(IllegalArgumentException::class) {
+                Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                    updateWhenMatched {
+                        it[Dest.value] = Source.value
+                    }
+                    updateWhenMatched {
+                        it[Dest.value] = Source.value
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun testDeleteInOracleAndSqlserver() {
+        withMergeTestsTables(excludeSettings = allDbExcept(TestDB.oracleRelatedDB + TestDB.sqlServerRelatedDB)) {
+            assertFailsWith(IllegalArgumentException::class) {
+                Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                    deleteWhenMatched()
+                    deleteWhenMatched()
+                }
+            }
+        }
+    }
+
+    @Test
+    fun testDeleteWhereCondition() {
+        withMergeTestsTables(excludeSettings = TestDB.oracleRelatedDB) {
+            assertFailsWith(IllegalArgumentException::class) {
+                Dest.mergeFrom(Source, on = { Source.key eq Dest.key }) {
+                    updateWhenMatched(deleteWhere = (Dest.value greater 1)) {
+                        it[Dest.value] = Source.value
+                    }
+                }
+            }
+        }
+    }
+
+    object SourceNoAutoId : IdTable<Int>("test_source_no_auto_id") {
+        override val id = integer("id").entityId()
+        val value = varchar("value", 128)
+        override val primaryKey = PrimaryKey(id)
+    }
```

Comment:
What's the reason for testing with an `IdTable` like this and not just a regular `Table` with a primary key?

Reply:
It was on purpose, I wanted to have table without id auto generation, to put the values I want and check that they are presented after merge (to test that ON condition is generated correctly)